### PR TITLE
[bugfix] on issue-label-change via API trigger webhooks too

### LINKED ISSUE #9403: Webhook not activated when setting a label on an issue
- Gitea version (or commit ref): 1.10.1
- Git version: 2.20.1
- Operating system: Raspbian GNU/Linux 10 (buster)
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x ] Not relevant
- Log gist:

## Description
When setting a label with a 3d party app (gitnex in my case) the corresponding event is not fired and the hook doesn't activate itself.

Step to reproduce :
- create a web hook reacting to issues events, and listen to it someway,
- create issue (API or web ui is ok : creating the ticket emits the event), 
- use the API to set a label to the issue : the label is correctly set in gitea, but the webhook is not activated
...

### PR DESCRIPTION
**Pull request by 6543**
close #9403

### CODE REVIEW

File: services/issue/label.go [hunk 1]
```diff
@@ -61,3 +61,13 @@ func RemoveLabel(issue *models.Issue, doer *models.User, label *models.Label) er
...
 	return nil
 }
+
+// ReplaceLabels removes all current labels and add new labels to the issue.
+func ReplaceLabels(issue *models.Issue, doer *models.User, labels []*models.Label) error {
+	if err := issue.ReplaceLabels(labels, doer); err != nil {
+		return err
+	}
+
+	notification.NotifyIssueChangeLabels(doer, issue, labels, nil)
```
Comment (guillep2k) [inline services/issue/label.go]: Maybe we should be notifying on deleted labels as well, and refrain to include unchanged labels.

Comment (6543) [reply services/issue/label.go]: @guillep2k done

**Comment by 6543** [comment]
@techknowlogick think this is an bug witch is already in 1.11.x

**Comment by 6543** [comment]
ci passed

**Comment by 6543** [comment]
ready to review

**Review by guillep2k — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by 6543** [comment]
what is wrong with this CI ...

**Comment by lafriks** [comment]
Complains about duplicate code:
routers/api/v1/repo/issue_label.go:70: 70-142 lines are duplicate of `routers/api/v1/repo/issue_label.go:216-288` (dupl)

**Comment by 6543** [comment]
now It's a refactor too ;)

**Comment by 6543** [comment]
wake up lgtm

File: routers/api/v1/repo/issue_label.go [hunk 2]
```diff
@@ -339,3 +307,28 @@ func ClearIssueLabels(ctx *context.APIContext) {
...
+		return
+	}
+
+	labels, err = models.GetLabelsInRepoByIDs(ctx.Repo.Repository.ID, form.Labels)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, "GetLabelsInRepoByIDs", err)
+		return
+	}
+
+	if !ctx.Repo.CanWriteIssuesOrPulls(issue.IsPull) {
```
Comment (guillep2k) [inline routers/api/v1/repo/issue_label.go:328]: We should check as follows:
- User can't READ this issue?: return 404
- User can't WRITE this issue?: return 403
- Other errors: return accordingly

The READ check should be performed first thing in the function.

**Review by guillep2k — CHANGES_REQUESTED** [review]
Sorry, I missed this check

File: routers/api/v1/repo/issue_label.go (hunk 2 above)
Comment (6543) [reply routers/api/v1/repo/issue_label.go:328]: @guillep2k already here!
https://github.com/go-gitea/gitea/blob/4e912d2ed869809066c98cf4ba3675de8086c29d/routers/api/v1/api.go#L709
is definde in scope of 
https://github.com/go-gitea/gitea/blob/4e912d2ed869809066c98cf4ba3675de8086c29d/routers/api/v1/api.go#L697
witch enfocrce this check:
https://github.com/go-gitea/gitea/blob/4e912d2ed869809066c98cf4ba3675de8086c29d/routers/api/v1/api.go#L736
witch return 404 if permission is not granted
https://github.com/go-gitea/gitea/blob/4e912d2ed869809066c98cf4ba3675de8086c29d/routers/api/v1/api.go#L478

Comment (6543) [reply routers/api/v1/repo/issue_label.go:328]: beside this I think ths changes of this kind are outside of scope of this BUGFIX @guillep2k 

I only had to refactor the code a bit to pass the lint checker

**Review by guillep2k — APPROVED** [review]

**Comment by x-beaudet** [comment]
smells good ! thanx ! :-)

**Comment by lunny** [comment]
Please send backport.

**Comment by 6543** [comment]
I will tomorow, if someone like to send a backport, feel free to

**Comment by 6543** [comment]
done ->  #10439

**Pull request merged**